<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to periodicDNA • periodicDNA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to periodicDNA">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">periodicDNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction-to-periodicdna.html">Introduction to periodicDNA</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to periodicDNA</h1>
                        <h4 class="author">Jacques Serizay</h4>
            
            <h4 class="date">2020-05-04</h4>
      
      
      <div class="hidden name"><code>introduction-to-periodicdna.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Many oligonucleotides are periodically occurring in genomes across eukaryotes, and some are impacting the physical properties of DNA. For example, DNA bendability is modulated by 10-bp periodic occurrences of WW (W = A/T) dinucleotides. periodicDNA is an R package offering a set of functions to identify k-mer periodicity in sequences of interest such as regulatory elements. It also provides a method to generate continuous tracks of k-mer periodicity over genomic loci.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>periodicDNA can be installed from GitHub as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)</a>
<a class="sourceLine" id="cb1-2" title="2">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"js2264/periodicDNA"</span>)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(periodicDNA)</a></code></pre></div>
</div>
<div id="loading-periodicdna-and-other-useful-packages" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-periodicdna-and-other-useful-packages" class="anchor"></a>Loading periodicDNA and other useful packages</h1>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(periodicDNA)</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co">#&gt; Loading required package: Biostrings</span></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="co">#&gt; Loading required package: XVector</span></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">#&gt; Attaching package: 'Biostrings'</span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co">#&gt;     strsplit</span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</a>
<a class="sourceLine" id="cb2-10" title="10"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GenomicRanges)</a>
<a class="sourceLine" id="cb2-11" title="11"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</a></code></pre></div>
</div>
<div id="how-periodna-works" class="section level1">
<h1 class="hasAnchor">
<a href="#how-periodna-works" class="anchor"></a>How perioDNA works</h1>
<p>When running <code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code> function for a given k-mer (typically a dinucleotide) over a set of GRanges, R does the following steps:</p>
<ol style="list-style-type: decimal">
<li>It first finds all the possible motif occurrences over the GRanges;<br>
</li>
<li>It then find all the possible pairwise distances and generates a normalised histogram of the pairwise distances;<br>
</li>
<li>It performs a Fast Fourier Transform of the normalised histogram;</li>
<li>Finally, it outputs the power spectrum density of the k-mer in the provided GRanges.</li>
</ol>
</div>
<div id="quick-k-mer-periodicity" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-k-mer-periodicity" class="anchor"></a>Quick k-mer periodicity</h1>
<p>A genome sequence and a set of genomic loci must be defined. Let’s focus on the C. elegans genome for now, and more specifically around its ubiquitous TSSs.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(ce_proms)</a>
<a class="sourceLine" id="cb3-2" title="2">ce_seq &lt;-<span class="st"> </span><span class="kw">getSeq</span>(</a>
<a class="sourceLine" id="cb3-3" title="3">    BSgenome.Celegans.UCSC.ce11<span class="op">::</span>BSgenome.Celegans.UCSC.ce11</a>
<a class="sourceLine" id="cb3-4" title="4">)</a>
<a class="sourceLine" id="cb3-5" title="5">ubiq_TSSs &lt;-<span class="st"> </span>ce_proms[ce_proms<span class="op">$</span>which.tissues <span class="op">==</span><span class="st"> 'Ubiq.'</span>] <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb3-6" title="6"><span class="st">    </span><span class="kw"><a href="../reference/deconvolveBidirectionalPromoters.html">deconvolveBidirectionalPromoters</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-7" title="7"><span class="st">    </span><span class="kw"><a href="../reference/alignToTSS.html">alignToTSS</a></span>(<span class="dv">50</span>, <span class="dv">300</span>)</a>
<a class="sourceLine" id="cb3-8" title="8">ubiq_TT &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPeriodicity.html">getPeriodicity</a></span>(</a>
<a class="sourceLine" id="cb3-9" title="9">    ubiq_TSSs, </a>
<a class="sourceLine" id="cb3-10" title="10">    <span class="dt">genome =</span> ce_seq, </a>
<a class="sourceLine" id="cb3-11" title="11">    <span class="dt">motif =</span> <span class="st">'TT'</span>, </a>
<a class="sourceLine" id="cb3-12" title="12">    <span class="dt">cores =</span> <span class="dv">1</span></a>
<a class="sourceLine" id="cb3-13" title="13">)</a>
<a class="sourceLine" id="cb3-14" title="14"><span class="co">#&gt; - Mapping k-mers.</span></a>
<a class="sourceLine" id="cb3-15" title="15"><span class="co">#&gt; - 3668916 k-mers found.</span></a>
<a class="sourceLine" id="cb3-16" title="16"><span class="co">#&gt; - Calculating pairwise distances.</span></a>
<a class="sourceLine" id="cb3-17" title="17"><span class="co">#&gt; - Normalizing histogram vector.</span></a>
<a class="sourceLine" id="cb3-18" title="18"><span class="co">#&gt; - Applying Fast Fourier Transform to the vector of distances.</span></a>
<a class="sourceLine" id="cb3-19" title="19"><span class="co">#&gt; - Computing signal-to-noise ratios.</span></a>
<a class="sourceLine" id="cb3-20" title="20"><span class="kw"><a href="../reference/plotPeriodicityResults.html">plotPeriodicityResults</a></span>(ubiq_TT)</a>
<a class="sourceLine" id="cb3-21" title="21"><span class="co">#&gt; [[1]]</span></a></code></pre></div>
<p><img src="introduction-to-periodicdna_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="introduction-to-periodicdna_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="introduction-to-periodicdna_files/figure-html/unnamed-chunk-4-3.png" width="700"></p>
</div>
<div id="fold-power-increase-metric" class="section level1">
<h1 class="hasAnchor">
<a href="#fold-power-increase-metric" class="anchor"></a>Fold Power Increase metric</h1>
<p>The FPI (fold power increase) metric was originally described in Pich et al., Cell 2018. It allows to accurately estimate the periodicity strength of a k-mer of interest by modelling the background periodicity.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">fpi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getFPI.html">getFPI</a></span>(</a>
<a class="sourceLine" id="cb6-2" title="2">    ce_seq[ubiq_TSSs],</a>
<a class="sourceLine" id="cb6-3" title="3">    <span class="dt">motif =</span> <span class="st">'TT'</span>, </a>
<a class="sourceLine" id="cb6-4" title="4">    <span class="dt">parallel_shuffling =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb6-5" title="5">    <span class="dt">n_shuffle =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb6-6" title="6">)</a>
<a class="sourceLine" id="cb6-7" title="7"><span class="co">#&gt; - Calculating observed PSD</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co">#&gt; - Mapping k-mers.</span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="co">#&gt; - 3668916 k-mers found.</span></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="co">#&gt; - Calculating pairwise distances.</span></a>
<a class="sourceLine" id="cb6-11" title="11"><span class="co">#&gt; - Normalizing histogram vector.</span></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="co">#&gt; - Applying Fast Fourier Transform to the vector of distances.</span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="co">#&gt; - Computing signal-to-noise ratios.</span></a>
<a class="sourceLine" id="cb6-14" title="14"><span class="co">#&gt; &gt;&gt; Measured PSD @ 10bp is: 3.06609400342861e-06</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="co">#&gt; - Shuffling 1/3</span></a>
<a class="sourceLine" id="cb6-16" title="16"><span class="co">#&gt; - Shuffling 2/3</span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="co">#&gt; - Shuffling 3/3</span></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="kw"><a href="../reference/plotFPI.html">plotFPI</a></span>(fpi)</a></code></pre></div>
<p><img src="introduction-to-periodicdna_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div id="generating-periodicity-tracks-with-periodicdna" class="section level1">
<h1 class="hasAnchor">
<a href="#generating-periodicity-tracks-with-periodicdna" class="anchor"></a>Generating periodicity tracks with periodicDNA</h1>
<p>Rather than computing the periodicity of a chosen k-mer in a given set of sequences, <code><a href="../reference/generatePeriodicityTrack.html">generatePeriodicityTrack()</a></code> function computes a linear periodicity track covering loci of interest.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="kw"><a href="../reference/generatePeriodicityTrack.html">generatePeriodicityTrack</a></span>(</a>
<a class="sourceLine" id="cb7-2" title="2">    ce_seq,</a>
<a class="sourceLine" id="cb7-3" title="3">    <span class="dt">granges =</span> ce_proms[<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb7-4" title="4">    <span class="dt">MOTIF =</span> <span class="st">'TT'</span>,</a>
<a class="sourceLine" id="cb7-5" title="5">    <span class="dt">FREQ =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb7-6" title="6">    <span class="dt">PROCS =</span> <span class="dv">1</span></a>
<a class="sourceLine" id="cb7-7" title="7">)</a>
<a class="sourceLine" id="cb7-8" title="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="co">#&gt;     The genome has been divided in 100-bp bins</span></a>
<a class="sourceLine" id="cb7-10" title="10"><span class="co">#&gt;     with a sliding window of 2-bp.</span></a>
<a class="sourceLine" id="cb7-11" title="11"><span class="co">#&gt;     </span></a>
<a class="sourceLine" id="cb7-12" title="12"><span class="co">#&gt;     549 windows overlap the input </span></a>
<a class="sourceLine" id="cb7-13" title="13"><span class="co">#&gt;     loci (extended to 1000-bp).</span></a>
<a class="sourceLine" id="cb7-14" title="14"><span class="co">#&gt;     </span></a>
<a class="sourceLine" id="cb7-15" title="15"><span class="co">#&gt;     The mapping will be split among 1 jobs.</span></a>
<a class="sourceLine" id="cb7-16" title="16"><span class="co">#&gt;     </span></a>
<a class="sourceLine" id="cb7-17" title="17"><span class="co">#&gt;     Each job will analyse 549 bins.</span></a>
<a class="sourceLine" id="cb7-18" title="18"><span class="co">#&gt;     </span></a>
<a class="sourceLine" id="cb7-19" title="19"><span class="co">#&gt;     Now starting...</span></a>
<a class="sourceLine" id="cb7-20" title="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-21" title="21"><span class="co">#&gt; &gt;&gt; 1 bins computed /// TIME: 2020-05-04 17:22:23</span></a>
<a class="sourceLine" id="cb7-22" title="22"><span class="co">#&gt; &gt;&gt; 549 bins computed /// TIME: 2020-05-04 17:22:47</span></a>
<a class="sourceLine" id="cb7-23" title="23"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-24" title="24"><span class="co">#&gt;    SUCCESS: TT-periodicity_g-100^2_b-100^5.bw has been created!</span></a>
<a class="sourceLine" id="cb7-25" title="25"><span class="co">#&gt;    1098 bases covered by the generated track.</span></a>
<a class="sourceLine" id="cb7-26" title="26"><span class="co">#&gt; [1] TRUE</span></a></code></pre></div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="co">#&gt; R version 3.5.2 (2018-12-20)</span></a>
<a class="sourceLine" id="cb8-3" title="3"><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="co">#&gt; Running under: Ubuntu 18.04.2 LTS</span></a>
<a class="sourceLine" id="cb8-5" title="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-6" title="6"><span class="co">#&gt; Matrix products: default</span></a>
<a class="sourceLine" id="cb8-7" title="7"><span class="co">#&gt; BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1</span></a>
<a class="sourceLine" id="cb8-8" title="8"><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span></a>
<a class="sourceLine" id="cb8-9" title="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-10" title="10"><span class="co">#&gt; locale:</span></a>
<a class="sourceLine" id="cb8-11" title="11"><span class="co">#&gt;  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              </span></a>
<a class="sourceLine" id="cb8-12" title="12"><span class="co">#&gt;  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    </span></a>
<a class="sourceLine" id="cb8-13" title="13"><span class="co">#&gt;  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   </span></a>
<a class="sourceLine" id="cb8-14" title="14"><span class="co">#&gt;  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 </span></a>
<a class="sourceLine" id="cb8-15" title="15"><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></a>
<a class="sourceLine" id="cb8-16" title="16"><span class="co">#&gt; [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span></a>
<a class="sourceLine" id="cb8-17" title="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-18" title="18"><span class="co">#&gt; attached base packages:</span></a>
<a class="sourceLine" id="cb8-19" title="19"><span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span></a>
<a class="sourceLine" id="cb8-20" title="20"><span class="co">#&gt; [8] methods   base     </span></a>
<a class="sourceLine" id="cb8-21" title="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-22" title="22"><span class="co">#&gt; other attached packages:</span></a>
<a class="sourceLine" id="cb8-23" title="23"><span class="co">#&gt;  [1] periodicDNA_0.3.0    Biostrings_2.50.2    XVector_0.22.0      </span></a>
<a class="sourceLine" id="cb8-24" title="24"><span class="co">#&gt;  [4] ggplot2_3.1.1        magrittr_1.5         GenomicRanges_1.34.0</span></a>
<a class="sourceLine" id="cb8-25" title="25"><span class="co">#&gt;  [7] GenomeInfoDb_1.18.2  IRanges_2.16.0       S4Vectors_0.20.1    </span></a>
<a class="sourceLine" id="cb8-26" title="26"><span class="co">#&gt; [10] BiocGenerics_0.28.0 </span></a>
<a class="sourceLine" id="cb8-27" title="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-28" title="28"><span class="co">#&gt; loaded via a namespace (and not attached):</span></a>
<a class="sourceLine" id="cb8-29" title="29"><span class="co">#&gt;  [1] SummarizedExperiment_1.12.0       zoo_1.8-5                        </span></a>
<a class="sourceLine" id="cb8-30" title="30"><span class="co">#&gt;  [3] tidyselect_0.2.5                  xfun_0.5                         </span></a>
<a class="sourceLine" id="cb8-31" title="31"><span class="co">#&gt;  [5] purrr_0.3.2                       lattice_0.20-38                  </span></a>
<a class="sourceLine" id="cb8-32" title="32"><span class="co">#&gt;  [7] colorspace_1.4-1                  htmltools_0.3.6                  </span></a>
<a class="sourceLine" id="cb8-33" title="33"><span class="co">#&gt;  [9] rtracklayer_1.42.2                yaml_2.2.0                       </span></a>
<a class="sourceLine" id="cb8-34" title="34"><span class="co">#&gt; [11] XML_3.98-1.19                     rlang_0.4.2                      </span></a>
<a class="sourceLine" id="cb8-35" title="35"><span class="co">#&gt; [13] pkgdown_1.4.1                     pillar_1.3.1                     </span></a>
<a class="sourceLine" id="cb8-36" title="36"><span class="co">#&gt; [15] glue_1.3.1                        withr_2.1.2                      </span></a>
<a class="sourceLine" id="cb8-37" title="37"><span class="co">#&gt; [17] BiocParallel_1.16.6               matrixStats_0.54.0               </span></a>
<a class="sourceLine" id="cb8-38" title="38"><span class="co">#&gt; [19] GenomeInfoDbData_1.2.0            plyr_1.8.4                       </span></a>
<a class="sourceLine" id="cb8-39" title="39"><span class="co">#&gt; [21] stringr_1.4.0                     zlibbioc_1.28.0                  </span></a>
<a class="sourceLine" id="cb8-40" title="40"><span class="co">#&gt; [23] munsell_0.5.0                     gtable_0.3.0                     </span></a>
<a class="sourceLine" id="cb8-41" title="41"><span class="co">#&gt; [25] memoise_1.1.0                     evaluate_0.13                    </span></a>
<a class="sourceLine" id="cb8-42" title="42"><span class="co">#&gt; [27] labeling_0.3                      Biobase_2.42.0                   </span></a>
<a class="sourceLine" id="cb8-43" title="43"><span class="co">#&gt; [29] knitr_1.22                        Rcpp_1.0.1                       </span></a>
<a class="sourceLine" id="cb8-44" title="44"><span class="co">#&gt; [31] BSgenome_1.50.0                   scales_1.0.0                     </span></a>
<a class="sourceLine" id="cb8-45" title="45"><span class="co">#&gt; [33] backports_1.1.3                   DelayedArray_0.8.0               </span></a>
<a class="sourceLine" id="cb8-46" title="46"><span class="co">#&gt; [35] desc_1.2.0                        fs_1.3.1                         </span></a>
<a class="sourceLine" id="cb8-47" title="47"><span class="co">#&gt; [37] Rsamtools_1.34.1                  digest_0.6.18                    </span></a>
<a class="sourceLine" id="cb8-48" title="48"><span class="co">#&gt; [39] stringi_1.3.1                     dplyr_0.8.0.1                    </span></a>
<a class="sourceLine" id="cb8-49" title="49"><span class="co">#&gt; [41] grid_3.5.2                        rprojroot_1.3-2                  </span></a>
<a class="sourceLine" id="cb8-50" title="50"><span class="co">#&gt; [43] tools_3.5.2                       bitops_1.0-6                     </span></a>
<a class="sourceLine" id="cb8-51" title="51"><span class="co">#&gt; [45] lazyeval_0.2.2                    RCurl_1.95-4.12                  </span></a>
<a class="sourceLine" id="cb8-52" title="52"><span class="co">#&gt; [47] tibble_2.1.1                      crayon_1.3.4                     </span></a>
<a class="sourceLine" id="cb8-53" title="53"><span class="co">#&gt; [49] pkgconfig_2.0.2                   BSgenome.Celegans.UCSC.ce11_1.4.2</span></a>
<a class="sourceLine" id="cb8-54" title="54"><span class="co">#&gt; [51] Matrix_1.2-17                     MASS_7.3-51.1                    </span></a>
<a class="sourceLine" id="cb8-55" title="55"><span class="co">#&gt; [53] assertthat_0.2.1                  rmarkdown_1.12.6                 </span></a>
<a class="sourceLine" id="cb8-56" title="56"><span class="co">#&gt; [55] R6_2.4.0                          GenomicAlignments_1.18.1         </span></a>
<a class="sourceLine" id="cb8-57" title="57"><span class="co">#&gt; [57] compiler_3.5.2</span></a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#loading-periodicdna-and-other-useful-packages">Loading periodicDNA and other useful packages</a></li>
      <li><a href="#how-periodna-works">How perioDNA works</a></li>
      <li><a href="#quick-k-mer-periodicity">Quick k-mer periodicity</a></li>
      <li><a href="#fold-power-increase-metric">Fold Power Increase metric</a></li>
      <li><a href="#generating-periodicity-tracks-with-periodicdna">Generating periodicity tracks with periodicDNA</a></li>
      <li><a href="#session-info">Session info</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://https://js2264.github.io/">Jacques Serizay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
